client/core/exchangeratefetcher.go

```diff
@@ -0,0 +1,270 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/dex"
+)
+
+const (
+	// DefaultFiatCurrency is the currency for displaying assets fiat value.
+	DefaultFiatCurrency = "USD"
+	// fiatRateRequestInterval is the amount of time between calls to the exchange API.
+	fiatRateRequestInterval = 5 * time.Minute
+	// fiatRateDataExpiry : Any data older than fiatRateDataExpiry will be discarded.
+	fiatRateDataExpiry = 60 * time.Minute
+
+	// Tokens. Used to identify fiat rate source.
+	messari       = "Messari"
+	coinpaprika   = "Coinpaprika"
+	dcrdataDotOrg = "dcrdata"
+)
+
+var (
+	dcrDataURL     = "https://explorer.dcrdata.org/api/exchangerate"
+	coinpaprikaURL = "https://api.coinpaprika.com/v1/tickers/%s"
+	messariURL     = "https://data.messari.io/api/v1/assets/%s/metrics/market-data"
+	btcBipId, _    = dex.BipSymbolID("btc")
+	dcrBipId, _    = dex.BipSymbolID("dcr")
+)
+
+// fiatRateFetchers is the list of all supported fiat rate fetchers.
+var fiatRateFetchers = map[string]rateFetcher{
+	coinpaprika:   fetchCoinpaprikaRates,
+	dcrdataDotOrg: fetchDcrdataRates,
+	messari:       fetchMessariRates,
+}
+
+// fiatRateInfo holds the fiat rate and the last update time for an
+// asset.
+type fiatRateInfo struct {
+	rate       float64
+	lastUpdate time.Time
+}
+
+// rateFetcher can fetch fiat rates for assets from an API.
+type rateFetcher func(context context.Context, logger dex.Logger, assets map[uint32]*SupportedAsset) map[uint32]float64
+
+type commonRateSource struct {
+	mtx         sync.RWMutex
```

Comment:
Should be clear from the struct definition what fields the mutex guards.  Please reformat so that the mutex is above all fields it protects and others are separate, as we do in other parts of the code.